packages/cli-lib/lib/walk.js

```diff
@@ -38,36 +44,52 @@ const generateFilePromise = (dir, file) => {
   });
 };
 
-const filesDataReducer = (allFiles, fileData) => {
-  switch (fileData.type) {
-    case STAT_TYPES.FILE:
-      return allFiles.concat(fileData.filepath);
-    case STAT_TYPES.DIRECTORY:
-      return allFiles.concat(fileData.files);
-    case STAT_TYPES.SYMBOLIC_LINK:
-      // Skip symlinks
-      return allFiles;
-    default:
-      return allFiles;
-  }
+const filesDataReducer = filesData => {
+  return filesData.reduce((acc, fileData) => {
+    switch (fileData.type) {
+      case STAT_TYPES.FILE:
+        return acc.concat(fileData.filepath);
+      case STAT_TYPES.DIRECTORY:
+        return acc.concat(fileData.files);
+      case STAT_TYPES.SYMBOLIC_LINK:
+        // Skip symlinks
+        return acc;
+      default:
+        return acc;
+    }
+  }, []);
 };
 
-function walk(dir) {
-  const processFiles = files => {
-    return Promise.all(files.map(file => generateFilePromise(dir, file)));
-  };
+function getFiles(dir, fileProcessor) {
   return new Promise((resolve, reject) => {
     fs.readdir(dir, (error, files) => {
       if (error) {
         reject(error);
       }
-      processFiles(files).then(filesData => {
-        resolve(filesData.reduce(filesDataReducer, []));
+      fileProcessor(files).then(filesData => {
+        resolve(filesDataReducer(filesData));
       });
     });
   });
 }
 
+function getDirectoryFiles(dir) {
```

Comment:
I updated this file per feedback I received [here](https://github.com/HubSpot/hubspot-cli/commit/7f1048dc081bc645e02cd21079842cf9b4d7e84e#r47510550). Tbh I'm not crazy about my solution, but my goal was to create two separate "deeply walk" and "shallow walk" functions for grabbing files in a folder that share logic. Let me know if you can think of a better way to handle this!

Reply:
More of a design question, but would make sense at all to walk the whole theme up front when the command is run and each validator could work off that file tree?

Reply:
Yeah I considered that, but decided against it b/c for the config validator I need to make sure that the `theme.json` file exists at the top level. Right now `walk` returns a flat list of all files so I wouldn't be able to determine whether the `theme.json` file was in the right place or not.

That being said, I'm guessing the `ThemeConfigValidator` is going to be the only validator that _only_ cares about top level files, and maybe running the walk command up front will still be an overall optimization.

Reply:
If we make the assumption that the directory being scanned is the root folder of a theme, one possible way of handling this would be to look for a file with a specific path in the flat listing (e.g. `files.filter(filepath => filepath === 'theme.json')`).

Reply:
Updated the code to only walk the files one time at the start of validation in 7da9192